system/lib/libc/crt1.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2019 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+// Much of this code comes from:
+// https://github.com/CraneStation/wasi-libc/blob/master/libc-bottom-half/crt/crt1.c
+// Converted malloc() calls to alloca() to avoid including malloc in all programs.
+
+#include <alloca.h>
+#include <stdlib.h>
+#include <sysexits.h>
+#include <wasi/wasi.h>
+#include <stdio.h>
+
+extern void __wasm_call_ctors(void) __attribute__((weak));
+extern int main(int argc, char** argv);
+
+void _start(void) {
+  /* Fill in the arguments from WASI syscalls. */
+  size_t argc;
+  char **argv;
+  __wasi_errno_t err;
+
+  /* Get the sizes of the arrays we'll have to create to copy in the args. */
+  size_t argv_buf_size;
+  err = __wasi_args_sizes_get(&argc, &argv_buf_size);
+  if (err != __WASI_ESUCCESS) {
+    __wasi_proc_exit(EX_OSERR);
+  }
+
+  if (argc) {
+    /* Allocate memory for the array of pointers, adding null terminator. */
+    argv = alloca(sizeof(char *) * (argc + 1));
+    /* Allocate memory for storing the argument chars. */
+    char *argv_buf = alloca(sizeof(char) * argv_buf_size);
+    /* Make sure the last pointer in the array is NULL. */
+    argv[argc] = NULL;
+    /* Fill the argument chars, and the argv array with pointers into those chars. */
+    err = __wasi_args_get(argv, argv_buf);
+    if (err != __WASI_ESUCCESS) {
+      __wasi_proc_exit(EX_OSERR);
+    }
+  }
```

Comment:
I am slightly unhappy that this code will exist even if `main` receives no params. But I guess there's no easy way to optimize that?

Perhaps this can almost work, as binaryen can inline main and original_main. But the allocas have side effects...

Perhaps metadce could be cleverer and inline the JS imports, which will just be `return 0` in this case, hmm...

Anyhow, nbd.

Reply:
I kind of agree.  That is why I'm only using the method for STANDALONE_WASM right now.  This issue might block the wider use of WASI by emscripten, but we can worry about that went to try to move in that direction.